spirv_msl.cpp

```diff
@@ -6748,6 +7041,245 @@ string CompilerMSL::image_type_glsl(const SPIRType &type, uint32_t id)
 	return img_type_name;
 }
 
+void CompilerMSL::emit_subgroup_op(const Instruction &i)
+{
+	const uint32_t *ops = stream(i);
+	auto op = static_cast<Op>(i.op);
+
+	// Metal 2.0 is required. iOS only supports quad ops. macOS only supports
+	// broadcast and shuffle on 10.13 (2.0), with full support in 10.14 (2.1).
+	// Note that iOS makes no distinction between a quad-group and a subgroup;
+	// all subgroups are quad-groups there.
+	if (!msl_options.supports_msl_version(2))
+		SPIRV_CROSS_THROW("Subgroups are only supported in Metal 2.0 and up.");
+
+	if (msl_options.is_ios())
+	{
+		switch (op)
+		{
+		default:
+			SPIRV_CROSS_THROW("iOS only supports quad-group operations.");
+		case OpGroupNonUniformElect:
+		case OpGroupNonUniformBroadcast:
+		case OpGroupNonUniformShuffle:
+		case OpGroupNonUniformShuffleXor:
+		case OpGroupNonUniformShuffleUp:
+		case OpGroupNonUniformShuffleDown:
+		case OpGroupNonUniformQuadSwap:
+		case OpGroupNonUniformQuadBroadcast:
+			break;
+		}
+	}
+
+	if (msl_options.is_macos() && !msl_options.supports_msl_version(2, 1))
+	{
+		switch (op)
+		{
+		default:
+			SPIRV_CROSS_THROW("Subgroup ops beyond broadcast and shuffle on macOS require Metal 2.0 and up.");
+		case OpGroupNonUniformElect:
+		case OpGroupNonUniformBroadcast:
+		case OpGroupNonUniformShuffle:
+		case OpGroupNonUniformShuffleXor:
+		case OpGroupNonUniformShuffleUp:
+		case OpGroupNonUniformShuffleDown:
+			break;
+		}
+	}
+
+	uint32_t result_type = ops[0];
+	uint32_t id = ops[1];
+
+	auto scope = static_cast<Scope>(get<SPIRConstant>(ops[2]).scalar());
+	if (scope != ScopeSubgroup)
+		SPIRV_CROSS_THROW("Only subgroup scope is supported.");
+
+	switch (op)
+	{
+	case OpGroupNonUniformElect:
+		// Vulkan spec says we have to support this if we support subgroups at all.
+		// But Metal prior to macOS 10.14 doesn't have the simd_is_first() function, and
+		// iOS doesn't have it at all. So we fake it by comparing the subgroup-local
+		// ID to 0. This isn't quite correct: this is supposed to return if we're the
+		// lowest *active* thread, but we'll otherwise be unable to support subgroups
+		// on macOS 10.13 or iOS.
+		if (msl_options.is_macos() && msl_options.supports_msl_version(2, 1))
```

Comment:
Can't we implement the elect on older targets in terms of simd_broadcast_first(subgroup_id) == subgroup_id?

Reply:
Unfortunately, no. Neither `simd_broadcast_first()` nor `quad_broadcast_first()` is available on iOS or macOS 10.13. I guess we won't be able to enable subgroup support for these devices.